src/hashutils.rs

```diff
@@ -0,0 +1,30 @@
+
+use ring::digest::{ Algorithm, Context };
+
+/// The sole purpose of this trait is to extend the standard
+/// `ring::algo::Algorithm` type with a couple utility functions.
+pub trait HashUtils {
+
+    /// Compute the hash the given byte array
+    fn hash_bytes(&'static self, bytes: &[u8]) -> Vec<u8>;
+
+    /// Compute the hash of the concatenation of `left` and `right`
+    fn combine_hashes(&'static self, left: &[u8], right: &[u8]) -> Vec<u8>;
+}
+
+impl HashUtils for Algorithm {
+
+    fn hash_bytes(&'static self, bytes: &[u8]) -> Vec<u8> {
+        let mut context = Context::new(self);
+        context.update(bytes);
+        context.finish().as_ref().into()
```

Comment:
In fact, you don't need to convert to `Vec<u8>`. You can return the `Digest` directly from this and the other function, saving a malloc each.